### PR TITLE
89 configure azure

### LINKED ISSUE #89: Configuring Azure
## Description
As a stakeholder I want to be able to access the latest stable version of the website to give feedback continuously to the developer team.

## Acceptance criteria
- [x] The website is publicly accessible
- [x] The website runs just like on our local machines

### PR DESCRIPTION
**Pull request by jonathan-lindqvist**
## Summary
The app is deployed and functions as it does locally.
Anyone can access the latest stable version; https://prodigoportal.azurewebsites.net/contracts

## Definition of done
- [x] All acceptance criteria are completed.
- [x] The code has (at least) 80% branch test coverage.
- [x] All tests pass.
- [ ] The code is reviewed and accepted by (at least) two other team members.
- [x] Views adhere to design style.

Closes #89

### CODE REVIEW

File: src/Infrastructure/EnvironmentVariableKeys.cs [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+
+/// <summary>
+/// All environment variable keys.
+/// </summary>
+public static class EnvironmentVariableKeys
+{
+    /// <summary>
+    /// The name of environment variable.
+    /// </summary>
+    public const string DbConnectionString = "prodigo_portal_db_connectionstring";
```
Comment (ellsib) [inline src/Infrastructure/EnvironmentVariableKeys.cs:11]: Shouldn't the connection to the database be hidden in some way?

File: src/Infrastructure/EFContractRepositoryFactory.cs [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+/// </summary>
+[ExcludeFromCodeCoverage]
+public class EFContractRepositoryFactory : IDesignTimeDbContextFactory<EFContractRepository>
+{
+    /// <exception cref="ArgumentException">When environment variable does not exist.</exception>
+    /// <inheritdoc />
+    public EFContractRepository CreateDbContext(string[] args)
+    {
+        var optionsBuilder = new DbContextOptionsBuilder<EFContractRepository>();
+        string? dbConnectionstring = Environment.GetEnvironmentVariable(EnvironmentVariableKeys.DbConnectionString);
```
Comment (ellsib) [inline src/Infrastructure/EFContractRepositoryFactory.cs:22]: I'm not really sure what an environment variable is. Is this common knowledge or should it be documented?

File: src/Infrastructure/Migrations/20220429145123_InitialCreate.Designer.cs [hunk 3]
```diff
@@ -3,75 +3,75 @@
...
 
 #nullable disable
 
-namespace Infrastructure.Migrations.PostgresContractRepositoryMigrations
+namespace Infrastructure.Migrations
 {
     [DbContext(typeof(EFContractRepository))]
-    [Migration("20220427114923_InitialContracts")]
-    partial class InitialContracts
+    [Migration("20220429145123_InitialCreate")]
```
Comment (ellsib) [inline src/Infrastructure/Migrations/20220429145123_InitialCreate.Designer.cs:15]: how do I create a new migration if needed later?

File: src/Infrastructure/Migrations/20220429145123_InitialCreate.Designer.cs [hunk 4]
```diff
@@ -3,75 +3,75 @@
...
                 {
                     b.Property<Guid>("Id")
                         .ValueGeneratedOnAdd()
-                        .HasColumnType("uuid");
+                        .HasColumnType("uniqueidentifier");
 
                     b.Property<string>("AdditionalDocument")
                         .IsRequired()
-                        .HasColumnType("text");
+                        .HasColumnType("nvarchar(max)");
```
Comment (ellsib) [inline src/Infrastructure/Migrations/20220429145123_InitialCreate.Designer.cs:35]: what is the difference between "text" and "nvarchar(max)"? Does it have to do with the change from Postgres?

**Review by ellsib — APPROVED** [review]
Looks good overall from what I can see. Left some comments, mostly for clarification. A bit tricky to understand all parts because I haven't seen it before. Also, the link in the description took a couple of minutes for me to open. That might be just for me, but can be good to check before merging.

File: src/Infrastructure/EnvironmentVariableKeys.cs (hunk 1 above)
Comment (martinjonsson01) [reply src/Infrastructure/EnvironmentVariableKeys.cs:11]: This is not the connection, this is the *name* of the environment variable that actually contains the connection string. So this is basically a fancy label for a hidden variable, but the label itself is not secret at all. What's very secret is the contents of this variable, but that's nowhere in the code.

File: src/Infrastructure/EFContractRepositoryFactory.cs (hunk 2 above)
Comment (martinjonsson01) [reply src/Infrastructure/EFContractRepositoryFactory.cs:22]: Any power user (or sysadmin, or developer) would be familiar with environment variables, as they're a common part of all operating systems. As a developer you often modify these when installing new development tools, for example.

File: src/Infrastructure/Migrations/20220429145123_InitialCreate.Designer.cs (hunk 3 above)
Comment (martinjonsson01) [reply src/Infrastructure/Migrations/20220429145123_InitialCreate.Designer.cs:15]: [MS Docs](https://docs.microsoft.com/en-us/ef/core/managing-schemas/migrations/?tabs=dotnet-core-cli) to the rescue! :)

File: src/Infrastructure/Migrations/20220429145123_InitialCreate.Designer.cs (hunk 4 above)
Comment (martinjonsson01) [reply src/Infrastructure/Migrations/20220429145123_InitialCreate.Designer.cs:35]: Yup, exactly. In Postgres strings would be stored as `text`, but in Microsoft SQL Server you'd store it as an `nvarchar`.

**Comment by martinjonsson01** [comment]
@ellsib 
> About the link: will this always get the latest version of the website (everything merged on Github)? It doesn't seem to be the case today, for example our delete button on users doesn't seem to exist :')

Yes, whenever we merge into master, that new code will be compiled into release mode and then uploaded to Azure. The reason for why the current version isn't the actual latest one is because we manually uploaded an older version when we were testing a few things. The plan is to not do manual uploads in the future though, so that shouldn't be an issue.

File: src/Infrastructure/EnvironmentVariableKeys.cs (hunk 1 above)
Comment (ellsib) [reply src/Infrastructure/EnvironmentVariableKeys.cs:11]: Alright! Thanks for the great explanations to all my comments :D

**Review by JacobBredin — APPROVED** [review]
LGTM! Hope the problems with the tests can be sorted out  🚀

I think Microsoft might have some problems with the image we are using: http://mcr.microsoft.com/mssql/server:2019-latest

**Comment by JacobBredin** [comment]
Running the tests locally now works 👍

**Pull request merged**